src/t_stream.c

```diff
@@ -1350,6 +1367,155 @@ robj *createObjectFromStreamID(streamID *id) {
                         id->ms,id->seq));
 }
 
+/* Returns non-zero if the ID is 0-0. */
+int streamIDEqZero(streamID *id) {
+    return !(id->ms || id->seq);
+}
+
+/* A helper that returns non-zero if the range from 'start' to the stream's
+ * end doesn't contain a tombstone.
+ *
+ * NOTE: this assumes that the caller had verified that 'start' is less than
+ * 's->last_id'. */
+int streamIsContiguousRange(stream *s, streamID *start) {
+    streamID start_id;
+
+    if (!s->length || streamIDEqZero(&s->xdel_max_id)) {
+        /* An empty stream or no tombstones. */
+        return 1;
+    }
+
+    /* Copy start ID, if given, or default to 0-0. */
+    if (start) {
+        start_id = *start;
+    } else {
+        start_id.ms = 0;
+        start_id.seq = 0;
+    }
+
+    if (streamCompareID(&s->first_id,&s->xdel_max_id) > 0) {
+        /* The latest tombstone is before the first entry. */
+        return 1;
+    }
+
+    if (streamCompareID(&start_id,&s->xdel_max_id) >= 0) {
+        /* The range doesn't include a tombstone. */
+        return 1;
+    }
+
+    /* The range includes a tombstone. */
+    return 0;
+}
+
+/* Replies with a consumerg group's current lag, that is the number of messages
+ * in the stream that are yet to be delivered. In case that the lag isn't
+ * available due to fragmentation, the reply to the client is a null. */
+void streamReplyWithCGLag(client *c, stream *s, streamCG *cg) {
+    int valid = 1;
+    uint64_t lag = 0;
+
+    if (!s->offset) {
+        /* The lag of a newly-initialized stream is 0. */
+        lag = 0;
+    } else if (cg->offset && streamIsContiguousRange(s,&cg->last_id)) {
+        /* No fragmentation ahead means that the group's offset is valid for
+         * performing the lag calculation. */
+        lag = s->offset - cg->offset;
+    } else if (streamIDEqZero(&cg->last_id)) {
+        if (streamIsContiguousRange(s,NULL)) {
+            /* The group is at 0-0 of a non-fragmented stream. */
+            lag = s->length;
+        } else {
+            valid = 0;
+        }
+    } else {
+        /* Attempt to retrieve the group's pffset. */
+        uint64_t offset = streamGetOffsetForTip(s,&cg->last_id);
+        if (offset) {
+            /* A valid offset was obtained. */
+            lag = s->offset - offset;
+        } else {
+            valid = 0;
+        }
+    }
+
+    if (valid) {
+        addReplyLongLong(c,(long long)lag);
+    } else {
+        addReplyNull(c);
+    }
+}
+
+/* A helper for getting an offset for the given ID if it is at one of the tips.
+ * A non-zero offset can be either logical (one before the first ID in the
+ * in the stream) or exact (the first or last IDs), but in both cases
+ * it provides a valid lag estimate. The zero offset usually means that
+ * the offset isn't available, except in the case of the 
+ * newly-initialized stream. */
+uint64_t streamGetOffsetForTip(stream *s, streamID *id) {
```

Comment:
maybe rename to `streamEstimateDistanceFromFirstEverEntry`

and please document the 4 return option (0, offset, s->offset - s->length, s->offset - s->length +1) and in which cases each is returned

Reply:
Please see the attempt to resolve via 5165321a4